cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -426,11 +439,69 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramId> programIds)
   }
 
 
-  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec)
+  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec,
+                                long creationTimeMillis, @Nullable String author, @Nullable String changeSummary)
     throws IOException {
-    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec)));
+    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec,
+                                                                                              changeSummary,
+                                                                                              creationTimeMillis,
+                                                                                              author))
+      , creationTimeMillis, author, changeSummary);
   }
 
+  public void createApplicationVersion(ApplicationId id, ApplicationMeta meta, @Nullable String parentVersion)
+    throws IOException, BadRequestException {
+    // Fetch the latest version
+    ApplicationMeta latest = getLatest(id.getNamespaceId(), id.getApplication()).stream().findAny().orElse(null);
+    //  If latest is null and app exists (legacy versions) then fetch -SNAPSHOT app
+    if (latest == null) {
+      latest = getApplication(id.getNamespace(), id.getApplication(), ApplicationId.DEFAULT_VERSION);
+    }
+    // If -SNAPSHOT version does not exist, return the app corresponding to the smallest version-id string
+    if (latest == null) {
+      latest = getAllAppVersions(id.getNamespace(), id.getApplication()).stream().findFirst().orElse(null);
+    }
+
+    if (!deployAppAllowed(parentVersion, latest)) {
+      throw new BadRequestException("Can't deploy the application because the parent version is not the " +
+                                                                    "latest.");
+    }
+    // When the app does not exist -it is not an edit).
+    if (latest != null) {
+      List<Field<?>> fields = getApplicationPrimaryKeys(id.getNamespace(), id.getApplication(),
+                                                        latest.getSpec().getAppVersion());
+      fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.LATEST_FIELD, "false"));
+      getApplicationSpecificationTable().upsert(fields);
```

Comment:
This should call `update` api since we made sure the row exists -- also, upserting a new row with only primary keys + `latest` field is not correct

Reply:
second thought.. after reading the `upsert` api, it made sure the untouched columns remain as is... So this should be fine. Ignore this comment